### DISCUSSION TITLE
Delaying the generating a static page until an asynchronous function is resolved

### DISCUSSION THREAD

**Discussion by PrinOrange** [vercel/next.js#71464] [Help]
### Summary

I am facing an issue with generating a static page in Next.js. For simplicity, let me explain with an example.

I want to render a component **only after** an asynchronous function has resolved, like this:

```js
const Page = ({ /* other props */ }) => {
  // The `fetchData` is an asynchronous function
  // I want the page to be generated only after fetchData is resolved  
  const data = await fetchData();

  return (<p>{data.message}</p>);
};

export async function getStaticProps() {
  /*other codes*/
  return { props: { /*other props*/ } };
}
export default Page;
```

Initially, I considered handling this within `getStaticProps` or `getServerSideProps`:

```js
export async function getStaticProps() {
  const data = await fetchData();
  return { props: { data, /*other props*/ } };
}
```

However, I encountered a limitation where the `data` returned by `fetchData` **cannot** be fully serialized into JSON and passed to the page via `props`. Hence, fetching the data on the server and passing it through props doesn't work in this case.

I also explored using `useEffect` and `useState` on the client side to fetch the data asynchronously:

```js
const Page = ({ /* other props */ }) => {
  const [data, setData] = useState();

  // However, useEffect only runs on the client-side, which affects SEO and first-page load
  useEffect(() => {
    fetchData().then((data) => setData(data));
  }, []);

  return (<p>{data?.message}</p>);
};

export async function getStaticProps() {
  /*other codes*/
  return { props: { /*other props*/ } };
}
export default Page;
```

But, since `useEffect` only runs on the client, this approach is not SEO-friendly and leads to delays in rendering for the first-page load.

#### Question

Is there any way to delay rendering a server component or static page in Next.js until an asynchronous function (defined within the component) resolves, without compromising on SEO or first-page performance?

Or, are there any new React or Next.js features that can help achieve this behavior more elegantly?

Or, does next.js have the concept of "static asynchronous pages"? In other words, static pages are asynchronous when they are built, and the page rendering results are output only after the page is built.

**Thank you very much**

**Comment by icyJoseph** [comment]
The question description doesn't make a whole lot of sense to be fair... could you draw a diagram of what you mean?

Taking the question title at face value, using App Router, you'd just need to do:

```jsx
const Page = async () => {
  const data = await fetchData();

  return (<p>{data.message}</p>);
};

export default Page;
```

I think maybe the confusion, is that you are using Pages Router, correct? 

Do you know exactly, 

>  the data returned by fetchData cannot be fully serialized into JSON and passed to the page via props

Which parts of the response are not allowing the serialization? What kind of data is it?
